pkgs/tools/virtualization/nixos-container/nixos-container.pl

```diff
@@ -214,29 +214,32 @@ sub writeNixOSConfig {
     die "$0: container ‘$containerName’ does not exist\n" ;
 }
 
+# Return the PID of the init process of the container.
+sub getLeader {
+    my $s = `machinectl show "$containerName" -p Leader`;
+    chomp $s;
+    $s =~ /^Leader=(\d+)$/ or die "unable to get container's main PID\n";
+    return int($1);
+}
+
 sub isContainerRunning {
     my $status = `systemctl show 'container\@$containerName'`;
     return $status =~ /ActiveState=active/;
 }
 
 sub terminateContainer {
+    my $leader = getLeader;
     system("machinectl", "terminate", $containerName) == 0
         or die "$0: failed to terminate container\n";
+    # Wait for the leader process to exit
+    while ( kill 0, $leader ) { sleep 1 }
```

Comment:
`sleep 0.1` would probably be preferred, as it is less noticeable.

This solution is unsafe because leader PID may be freed and reused by an unrelated process while `terminateContainer` sleeps. `machinectl` (or `systemctl`) should be used for querying the status.

Reply:
> sleep 0.1 would probably be preferred, as it is less noticeable.

@orivej I agree and I'd also prefer a shorter duration, but as far as I understand [from here](http://perldoc.perl.org/functions/sleep.html), in perl you can't sleep less than 1 second, unless you use the `Time::HiRes` module, which I don't know if it's available here. Improvements welcome.






> This solution is unsafe because leader PID may be freed and reused by an unrelated process while terminateContainer sleeps.

Yes

> machinectl (or systemctl) should be used for querying the status.

@orivej Do you know if there's a UUID or something like that provided by `machinectl`, or do we have to wait based on the name?

Reply:
> as far as I understand [from here](http://perldoc.perl.org/functions/sleep.html), in perl you can't sleep less than 1 second, unless you use the `Time::HiRes` module, which I don't know if it's available here. Improvements welcome.

That page says that this module is shipped with Perl 5.8 and later, so you can easily use it.

> Do you know if there's a UUID or something like that provided by `machinectl`, or do we have to wait based on the name?

I don't know, I have not looked at `machinectl` before this PR.

Reply:
I suggest we add a comment to improve reliability, this is an improvement on it's own compared to current implementation.

Reply:
I have swithed to `Time::HiRes` as @orivej suggested, and added a TODO about switching away from PIDs in `nixos-container`.

Reply:
I've filed a separate issue to track the removal of PID based logic, and found that `machinectl show` already provides an `Id` field we could use: https://github.com/NixOS/nixpkgs/issues/33171